targets/TARGET_STM/sleep.c

```diff
@@ -40,17 +40,25 @@ extern void HAL_ResumeTick(void);
 
 void hal_sleep(void)
 {
+    // Disable IRQs
+    core_util_critical_section_enter();
```

Comment:
In my experience (although that's mainly on ARM-A), any WFI-type sleep function should always be called with interrupts disabled. 

Otherwise there's a race condition - you may have been idle, which is why you were going into sleep, but as you start running the sleep routine, an interrupt wakes a thread, but then sleep goes ahead and does WFI, so we block until the next interrupt, and the woken thread doesn't run immediately.

If interrupts are disabled all the way from the "are we idle" check to the "WFI", then the sleep routine will correctly return immediately due to a pending interrupt.

So I think 0xc0170 is right - hal_sleep() should probably always be entered with interrupts disabled.